File: bench/src/cs/src/csvagg.cs
using System;
using System.Diagnostics;

static class Bench
{
    const ulong ROWS = 400000;

    static readonly string[] REGIONS = { "EMEA", "APAC", "AMER", "LATAM", "NORDIC", "IBERIA", "BENELUX", "DACH" };

    static ulong gSeed = 12345;

    static ulong Rnd()
    {
        gSeed = (gSeed * 16807) % 2147483647;
        return gSeed;
    }

    sealed class ByteMap
    {
        public ulong[] KeyOff;
        public ulong[] KeyLen;
        public ulong[] Val;
        public byte[] Used;
        public byte[] Base;
        public ulong Mask;
        public ulong Count;

        public ByteMap(ulong capacity, byte[] b)
        {
            int c = (int) capacity;
            KeyOff = new ulong[c];
            KeyLen = new ulong[c];
            Val = new ulong[c];
            Used = new byte[c];
            Base = b;
            Mask = capacity - 1;
            Count = 0;
        }

        public ulong Probe(ulong off, ulong len)
        {
            ulong h = 2166136261;
            for (ulong i = 0; i < len; i++)
            {
                h ^= Base[(int) (off + i)];
                h = (h * 16777619) & 0xFFFFFFFF;
            }

            ulong idx = h & Mask;
            while (Used[(int) idx] != 0)
            {
                if (KeyLen[(int) idx] == len && MemCmp(Base, (int) KeyOff[(int) idx], (int) off, (int) len) == 0)
                    return idx;
                idx = (idx + 1) & Mask;
            }

            return idx;
        }

        public void Claim(ulong idx, ulong off, ulong len, ulong value)
        {
            int i = (int) idx;
            Used[i] = 1;
            KeyOff[i] = off;
            KeyLen[i] = len;
            Val[i] = value;
            Count += 1;
        }
    }

    static int MemCmp(byte[] b, int a, int c, int n)
    {
        for (int i = 0; i < n; i++)
        {
            if (b[a + i] != b[c + i])
                return b[a + i] < b[c + i] ? -1 : 1;
        }
        return 0;
    }

    static ulong WriteUInt(byte[] buf, ulong pos, ulong v)
    {
        var tmp = new byte[24];
        ulong n = 0;
        ulong x = v;
        if (x == 0)
        {
            tmp[0] = (byte) '0';
            n = 1;
        }
        while (x > 0)
        {
            tmp[(int) n] = (byte) (48 + (x % 10));
            n += 1;
            x /= 10;
        }

        ulong p = pos;
        ulong i = n;
        while (i > 0)
        {
            i -= 1;
            buf[(int) p] = tmp[(int) i];
            p += 1;
        }
        return p;
    }

    static ulong WriteUInt2(byte[] buf, ulong pos, ulong v)
    {
        buf[(int) pos] = (byte) (48 + (v / 10));
        buf[(int) pos + 1] = (byte) (48 + (v % 10));
        return pos + 2;
    }

    static int Main()
    {
        // ---- data generation (not timed) ----
        var text = new byte[ROWS * 48];
        ulong n = 0;

        for (ulong j = 0; j < ROWS; j++)
        {
            if (j > 0)
            {
                text[(int) n] = (byte) '\n';
                n += 1;
            }

            string region = REGIONS[(int) (Rnd() % 8)];
            ulong y = 2024 + (Rnd() % 3);
            ulong m = 1 + (Rnd() % 12);
            ulong d = 1 + (Rnd() % 28);
            ulong qty = 1 + (Rnd() % 50);
            ulong cents = 100 + (Rnd() % 99900);

            n = WriteUInt(text, n, j);
            text[(int) n] = (byte) ',';
            n += 1;
            for (int i = 0; i < region.Length; i++)
            {
                text[(int) n] = (byte) region[i];
                n += 1;
            }
            text[(int) n] = (byte) ',';
            n += 1;
            n = WriteUInt(text, n, y);
            text[(int) n] = (byte) '-';
            n += 1;
            n = WriteUInt2(text, n, m);
            text[(int) n] = (byte) '-';
            n += 1;
            n = WriteUInt2(text, n, d);
            text[(int) n] = (byte) ',';
            n += 1;
            n = WriteUInt(text, n, qty);
            text[(int) n] = (byte) ',';
            n += 1;
            n = WriteUInt(text, n, cents / 100);
            text[(int) n] = (byte) '.';
            n += 1;
            n = WriteUInt2(text, n, cents % 100);
        }

        // ---- timed work ----
        long start_t = Stopwatch.GetTimestamp();

        var agg = new ByteMap(64, text);

        var slotCount = new ulong[64];
        var slotQty = new ulong[64];
        var slotRev = new double[64];
        var slotMax = new double[64];

        ulong pos = 0;
        ulong rows = 0;

        while (pos < n)
        {
            ulong eol = pos;
            while (eol < n && text[(int) eol] != (byte) '\n')
                eol += 1;

            // field 0: id
            ulong p = pos;
            while (p < eol && text[(int) p] != (byte) ',')
                p += 1;
            p += 1;

            // field 1: region
            ulong rs = p;
            while (p < eol && text[(int) p] != (byte) ',')
                p += 1;
            ulong rlen = p - rs;
            p += 1;

            // field 2: date (skipped)
            while (p < eol && text[(int) p] != (byte) ',')
                p += 1;
            p += 1;

            // field 3: qty
            ulong qty = 0;
            while (p < eol && text[(int) p] != (byte) ',')
            {
                qty = qty * 10 + (ulong) (text[(int) p] - 48);
                p += 1;
            }
            p += 1;

            // field 4: price
            ulong ip = 0;
            while (p < eol && text[(int) p] != (byte) '.')
            {
                ip = ip * 10 + (ulong) (text[(int) p] - 48);
                p += 1;
            }
            p += 1;
            ulong fr = 0;
            while (p < eol)
            {
                fr = fr * 10 + (ulong) (text[(int) p] - 48);
                p += 1;
            }
            double price = (double) ip + (double) fr / 100.0;

            ulong idx = agg.Probe(rs, rlen);
            if (agg.Used[(int) idx] == 0)
            {
                int slot = (int) agg.Count;
                agg.Claim(idx, rs, rlen, (ulong) slot);
                slotCount[slot] = 1;
                slotQty[slot] = qty;
                slotRev[slot] = (double) qty * price;
                slotMax[slot] = price;
            }
            else
            {
                int slot = (int) agg.Val[(int) idx];
                slotCount[slot] += 1;
                slotQty[slot] += qty;
                slotRev[slot] += (double) qty * price;
                if (price > slotMax[slot])
                    slotMax[slot] = price;
            }

            rows += 1;
            pos = eol + 1;
        }

        // sort the region keys, then fold them in order
        var order = new ulong[64];
        var keyIdx = new ulong[64];
        ulong nk = 0;
        for (int i = 0; i < 64; i++)
        {
            if (agg.Used[i] != 0)
            {
                keyIdx[(int) nk] = (ulong) i;
                nk += 1;
            }
        }

        for (int i = 0; i < (int) nk; i++)
            order[i] = (ulong) i;
        for (int i = 0; i < (int) nk; i++)
        {
            int best = i;
            for (int j = 0; j < (int) nk; j++)
            {
                if (j <= i)
                    continue;
                int a = (int) keyIdx[(int) order[best]];
                int b = (int) keyIdx[(int) order[j]];
                ulong la = agg.KeyLen[a];
                if (agg.KeyLen[b] < la)
                    la = agg.KeyLen[b];
                int c = MemCmp(text, (int) agg.KeyOff[b], (int) agg.KeyOff[a], (int) la);
                if (c < 0 || (c == 0 && agg.KeyLen[b] < agg.KeyLen[a]))
                    best = j;
            }
            ulong t = order[i];
            order[i] = order[best];
            order[best] = t;
        }

        ulong check = rows;
        for (int k = 0; k < (int) nk; k++)
        {
            int slot = (int) agg.Val[(int) keyIdx[(int) order[k]]];
            check += ((ulong) k + 1) * ((ulong) (slotRev[slot] * 100.0 + 0.5) % 1000003);
            check += slotCount[slot] + slotQty[slot] + (ulong) (slotMax[slot] * 100.0 + 0.5);
        }

        double ms = (Stopwatch.GetTimestamp() - start_t) * 1000.0 / Stopwatch.Frequency;
        Console.WriteLine($"CHECK={check} MS={ms:F6}");
        return 0;
    }
}

File: bench/src/rust/csvagg.rs
use std::time::Instant;

const ROWS: u64 = 400000;

const REGIONS: [&str; 8] = [
    "EMEA", "APAC", "AMER", "LATAM", "NORDIC", "IBERIA", "BENELUX", "DACH",
];

static mut G_SEED: u64 = 12345;

fn rnd() -> u64 {
    unsafe {
        G_SEED = (G_SEED * 16807) % 2147483647;
        G_SEED
    }
}

struct ByteMap {
    key_off: Vec<u64>,
    key_len: Vec<u64>,
    val: Vec<u64>,
    used: Vec<u8>,
    mask: u64,
    count: u64,
}

impl ByteMap {
    fn new(capacity: u64) -> ByteMap {
        let c = capacity as usize;
        ByteMap {
            key_off: vec![0; c],
            key_len: vec![0; c],
            val: vec![0; c],
            used: vec![0; c],
            mask: capacity - 1,
            count: 0,
        }
    }

    fn probe(&self, base: &[u8], off: u64, len: u64) -> u64 {
        let mut h: u64 = 2166136261;
        for i in 0..len {
            h ^= base[(off + i) as usize] as u64;
            h = (h.wrapping_mul(16777619)) & 0xFFFFFFFF;
        }

        let mut idx = h & self.mask;
        while self.used[idx as usize] != 0 {
            if self.key_len[idx as usize] == len {
                let a = self.key_off[idx as usize] as usize;
                let n = len as usize;
                if base[a..a + n] == base[off as usize..off as usize + n] {
                    return idx;
                }
            }
            idx = (idx + 1) & self.mask;
        }

        idx
    }

    fn claim(&mut self, idx: u64, off: u64, len: u64, value: u64) {
        let i = idx as usize;
        self.used[i] = 1;
        self.key_off[i] = off;
        self.key_len[i] = len;
        self.val[i] = value;
        self.count += 1;
    }
}

fn write_uint(buf: &mut [u8], pos: u64, v: u64) -> u64 {
    let mut tmp = [0u8; 24];
    let mut n: u64 = 0;
    let mut x = v;
    if x == 0 {
        tmp[0] = b'0';
        n = 1;
    }
    while x > 0 {
        tmp[n as usize] = (48 + (x % 10)) as u8;
        n += 1;
        x /= 10;
    }

    let mut p = pos;
    let mut i = n;
    while i > 0 {
        i -= 1;
        buf[p as usize] = tmp[i as usize];
        p += 1;
    }
    p
}

fn write_uint2(buf: &mut [u8], pos: u64, v: u64) -> u64 {
    buf[pos as usize] = (48 + (v / 10)) as u8;
    buf[(pos + 1) as usize] = (48 + (v % 10)) as u8;
    pos + 2
}

fn main() {
    // ---- data generation (not timed) ----
    let mut text = vec![0u8; (ROWS * 48) as usize];
    let mut n: u64 = 0;

    for j in 0..ROWS {
        if j > 0 {
            text[n as usize] = b'\n';
            n += 1;
        }

        let region = REGIONS[(rnd() % 8) as usize].as_bytes();
        let y = 2024 + (rnd() % 3);
        let m = 1 + (rnd() % 12);
        let d = 1 + (rnd() % 28);
        let qty = 1 + (rnd() % 50);
        let cents = 100 + (rnd() % 99900);

        n = write_uint(&mut text, n, j);
        text[n as usize] = b',';
        n += 1;
        for i in 0..region.len() {
            text[n as usize] = region[i];
            n += 1;
        }
        text[n as usize] = b',';
        n += 1;
        n = write_uint(&mut text, n, y);
        text[n as usize] = b'-';
        n += 1;
        n = write_uint2(&mut text, n, m);
        text[n as usize] = b'-';
        n += 1;
        n = write_uint2(&mut text, n, d);
        text[n as usize] = b',';
        n += 1;
        n = write_uint(&mut text, n, qty);
        text[n as usize] = b',';
        n += 1;
        n = write_uint(&mut text, n, cents / 100);
        text[n as usize] = b'.';
        n += 1;
        n = write_uint2(&mut text, n, cents % 100);
    }

    // ---- timed work ----
    let start_t = Instant::now();

    let mut agg = ByteMap::new(64);

    let mut slot_count = vec![0u64; 64];
    let mut slot_qty = vec![0u64; 64];
    let mut slot_rev = vec![0.0f64; 64];
    let mut slot_max = vec![0.0f64; 64];

    let mut pos: u64 = 0;
    let mut rows: u64 = 0;

    while pos < n {
        let mut eol = pos;
        while eol < n && text[eol as usize] != b'\n' {
            eol += 1;
        }

        // field 0: id
        let mut p = pos;
        while p < eol && text[p as usize] != b',' {
            p += 1;
        }
        p += 1;

        // field 1: region
        let rs = p;
        while p < eol && text[p as usize] != b',' {
            p += 1;
        }
        let rlen = p - rs;
        p += 1;

        // field 2: date (skipped)
        while p < eol && text[p as usize] != b',' {
            p += 1;
        }
        p += 1;

        // field 3: qty
        let mut qty: u64 = 0;
        while p < eol && text[p as usize] != b',' {
            qty = qty * 10 + (text[p as usize] - 48) as u64;
            p += 1;
        }
        p += 1;

        // field 4: price
        let mut ip: u64 = 0;
        while p < eol && text[p as usize] != b'.' {
            ip = ip * 10 + (text[p as usize] - 48) as u64;
            p += 1;
        }
        p += 1;
        let mut fr: u64 = 0;
        while p < eol {
            fr = fr * 10 + (text[p as usize] - 48) as u64;
            p += 1;
        }
        let price = ip as f64 + fr as f64 / 100.0;

        let idx = agg.probe(&text, rs, rlen);
        if agg.used[idx as usize] == 0 {
            let slot = agg.count as usize;
            agg.claim(idx, rs, rlen, slot as u64);
            slot_count[slot] = 1;
            slot_qty[slot] = qty;
            slot_rev[slot] = qty as f64 * price;
            slot_max[slot] = price;
        } else {
            let slot = agg.val[idx as usize] as usize;
            slot_count[slot] += 1;
            slot_qty[slot] += qty;
            slot_rev[slot] += qty as f64 * price;
            if price > slot_max[slot] {
                slot_max[slot] = price;
            }
        }

        rows += 1;
        pos = eol + 1;
    }

    // sort the region keys, then fold them in order
    let mut order = [0u64; 64];
    let mut key_idx = [0u64; 64];
    let mut nk: u64 = 0;
    for i in 0..64usize {
        if agg.used[i] != 0 {
            key_idx[nk as usize] = i as u64;
            nk += 1;
        }
    }

    for i in 0..nk as usize {
        order[i] = i as u64;
    }
    for i in 0..nk as usize {
        let mut best = i;
        for j in 0..nk as usize {
            if j <= i {
                continue;
            }
            let a = key_idx[order[best] as usize] as usize;
            let b = key_idx[order[j] as usize] as usize;
            let mut la = agg.key_len[a];
            if agg.key_len[b] < la {
                la = agg.key_len[b];
            }
            let la = la as usize;
            let sb = &text[agg.key_off[b] as usize..agg.key_off[b] as usize + la];
            let sa = &text[agg.key_off[a] as usize..agg.key_off[a] as usize + la];
            match sb.cmp(sa) {
                std::cmp::Ordering::Less => best = j,
                std::cmp::Ordering::Equal => {
                    if agg.key_len[b] < agg.key_len[a] {
                        best = j;
                    }
                }
                std::cmp::Ordering::Greater => {}
            }
        }
        order.swap(i, best);
    }

    let mut check = rows;
    for k in 0..nk as usize {
        let slot = agg.val[key_idx[order[k] as usize] as usize] as usize;
        check += (k as u64 + 1) * ((slot_rev[slot] * 100.0 + 0.5) as u64 % 1000003);
        check += slot_count[slot] + slot_qty[slot] + (slot_max[slot] * 100.0 + 0.5) as u64;
    }

    let ms = start_t.elapsed().as_secs_f64() * 1000.0;
    println!("CHECK={} MS={:.6}", check, ms);
}
